### ISSUE TITLE
.gitignore

### ISSUE THREAD

**Issue by vogler** [janestreet/jbuilder#207]
It would be nice to get away without a `.gitignore` file in git 😄 

We should generate a `_build/.gitignore` with `*` as content. This way the whole directory will be ignored.

Hopefully `.merlin` can be moved to `_build`. See https://github.com/ocaml/merlin/issues/686.

That only leaves `<package>.install`. I guess for this the user's `.gitignore` or `<package>.opam` would have to be modified? Or better, adjust opam?

**Comment by diml** [comment]
The `_build/.gitignore` seems good to me. For the `<package>.install`, yh it seems that opam would need to be updated

**Comment by vogler** [comment]
Is there something that speaks against a `_build/.gitignore`?
Using `.gitignore_global` for directories always seems like a big hack to me. People have to do it manually and therefore might first accidentally commit `_build` in their dune project, and then, afterwards, they might miss to commit a directory (or file) `_build` in some other project (ok, unlikely given the name, but could happen).

**Comment by diml** [comment]
Not against `_build/.gitignore` specifically, but it is indeed a bit odd to do it only for `_build` since we generate other files in the source tree, i.e. .merlin and .install files. There is a plan to stop generating .merlin files, and I suppose some work could be done on opam to look for these files in the _build directory. If this happens, then generating `_build/.gitignore` files seems fine.

**Comment by rgrinberg** [comment]
I'm against this feature for 3 reasons:

* It does not completely solve the problem that it purports to solve

* It does nothing for a good chunk of users. Many people use a global gitignore 

* Git is not the only version control system in use.

However, there's a few git specific features that people have requested from dune. I'm wondering if we should have a `git` plugin that people can enable in their project. This would give them access to these extra targets, the ability to substitute the current hash, etc.

**Comment by diml** [comment]
They way I see it, if most users have to write the same entry in their configuration files, then the tool is simply using the wrong default. In this case, I agree that it doesn't make much sense to generate `_build/.gitignore` because of .merlin and .install files, but if these are sorted out, then generating `_build/.gitignore` would allow allow many users to not have to customize their gitignore either locally or globally, which is an improvement.

**Comment by rgrinberg** [comment]
By the way, in the end, if someone makes a PR to add this feature, I wouldn't be against accepting it. The feature seems to have a few flaws and relatively small benefit, but I can't see how it might harm anyone. It's just that I don't think it's worth it for anyone the dune team or potential contributors to work on this.

**Comment by diml** [comment]
Yeah, it's definitely not a priority. That said I would be happy to have it. I didn't know about gitignore_global. I suppose it helps but there is something that slightly bothers me about it: the set of files to ignore basically depend on the class of project you are working on. For instance , maybe some other language uses .install files as source files and they shouldn't be ignored.

**Comment by ELLIOTTCABLE** [comment]
So, I'm not sure if I'm speaking out of turn here, but — at least theoretically — is there much difference between, say, `.merlin` files and `.gitignore` or `.hgignore` files? What about Opam integration?

The way I see it, if a tool is:

1. In use by a majority of the community,
2. Needs some sort of special support to remove obvious stumbling blocks in basically everyone's workflow,
2. And is relatively straightforward to add special support for,

… then it's worth supporting. Even if that support dosen't feel OCaml-specific, or feels like it's prioritizing one tool over another.

Relatedly: maybe instead of a bunch of separate codepaths for various systems — just have one set of as-simple-as-possible heuristics, aiming for lowest maintenance burden over perfection in detection, and generally erring on the side of assuming VCS is indeed present; then if *any* VCS is apparently present, enable *all* of the VCS-supporte features?
